syntax = "proto3";

package video_processing;

// данные изображения (без дублирования метаданных)
message ImageData {
    uint32 width = 1;
    uint32 height = 2;
    PixelFormat pixel_format = 3;
    ImageEncoding encoding = 4;
    bytes image_data = 5;
}

// Пара изображений (original + processed)
message ImagePair {
    ImageData original = 1;
    ImageData processed = 2;
}

// Контейнер передачи
message VideoFrame {
    uint64 frame_id = 1;
    double timestamp = 2; 
    string sender_id = 3;
    FrameType frame_type = 4;
    
    oneof content {
        ImageData single_image = 5;    // 1 изображение
        ImagePair image_pair = 6;      // 2 изображения
        string command = 7;            // 0 изображений (команда)
    }
}

enum FrameType {
    UNKNOWN = 0;
    CAPTURED_FRAME = 1;    // От Capturer к Worker
    PROCESSED_FRAME = 2;   // От Worker к Composer  
    CONTROL_MESSAGE = 3;   // Команда управления
}

enum PixelFormat {
    BGR = 0;
    RGB = 1;
    GRAY = 2;
}

enum ImageEncoding {
    JPEG = 0;
    PNG = 1;
    RAW = 2;
}
